File: lengthOfLongestSubstring.js
// Time Complexity : O(n) where n is length of string s
// Space Complexity : O(1) constant because at most only 26 chars
// Did this code successfully run on LeetCode : Yes
// Any problem you faced while coding this : No

/**
 * @param {string} s
 * @return {number}
 */

//sliding window sloution
var lengthOfLongestSubstring = function(s) {
    let charMap = new Map();
    let slow = 0;
    let max = 0;
    for(let i = 0; i < s.length; i++) {
        char = s.charAt(i);
        if(charMap.has(char)) {
            slow = Math.max(slow,charMap.get(char));
        }
        max = Math.max(max, i - slow + 1)
        charMap.set(char,i+1);
    }
    return max;
};

File: LengthOfLongestSubString.cs
using System;
using System.Collections.Generic;

namespace Algorithms
{
    public class LengthOfLongestSubString
    {
        /// Time Complexity : O(n) 
        // Space Complexity :O(n)
        // Did this code successfully run on Leetcode :Yes
        // Any problem you faced while coding this :  Referred Online
        public int LengthOfLongestSubstring(string s)
        {
            int maxLength = 0;

            Dictionary<char, int> map = new Dictionary<char, int>();
            for (int fast = 0, slow = 0; fast < s.ToCharArray().Length; fast++)
            {
                if (map.ContainsKey(s.ToCharArray()[fast]))
                {
                    slow = Math.Max(map.GetValueOrDefault(s.ToCharArray()[fast]), slow);
                }
                maxLength = Math.Max(maxLength, fast - slow + 1);
                map[s.ToCharArray()[fast]] = fast + 1;
            }
            return maxLength;
        }
    }
}
